![title](title.png)

# Лабораторна робота №1

## Тема

Підсумкові функції та відображення

## Мета

навчитися вибрати відповідні дані
з DataFrame або Series, вибирати правильні дані з набору даних.

## Автор

Молчанов Михайло Валерійович, ІА-12, Варіант 16

## Виконання



### Імпортуємо бібліотеки

Заімпортуємо бібіліотеки, щоб їх просто далі використовувати

In [2]:
import numpy as np
import pandas as pd

### Отримання дата сету

Дата сет для 16-го варіанту не вдалося завантажити, оскільки він був видалений, тому було завантажено датасет для 15-го
варіанту. 

In [18]:
# I placed data set in my data folder and added it to gitignore, so you can't see it, but it's there.
file_path = 'data/Cancer_rate_by_countries.csv'

# Let's load the data set.
cancer_data = pd.read_csv(file_path)

# And view some of its rows, so we have an idea what we are dealing with.
cancer_data.head()

Unnamed: 0  Rank        Country  Cancer_Rate
0           0     1      Australia        468.0
1           1     2    New Zealand        438.1
2           2     3        Ireland        373.7
3           3     4        Hungary        368.1
4           4     5  United States        352.2

### Аналіз дата сету

Дата сет було завантажено успішно, та тепер можна його детально проаналізувати. З першого вигляду видно, що у нас є наступні колонки:

- Rank - рейтинг країн за рівнем захворювання
- Country - назва країни
- Cancer_Rate - рівень захворювання раком

#### Describe
Проаналізуємо ці данні детальніше за допомогою методу `describe()`. 
Метод `describe()` в бібліотеці Pandas автоматично генерує підсумкові статистичні дані для числових стовпців у наборі даних. Він надає основні показники, такі як:

- Кількість непорожніх значень (count)
- Середнє значення (mean)
- Стандартне відхилення (std)
- Мінімальне (min) та максимальне значення (max)
- Квартилі (25%, 50%, 75%)

Цей метод є дуже корисним для швидкого огляду основних характеристик датасету та виявлення можливих аномалій або тенденцій.

In [6]:
# Загальний опис параметрів дата сету
cancer_data.describe()

Unnamed: 0      Rank  Cancer_Rate
count    50.00000  50.00000    50.000000
mean     24.50000  25.50000   294.082000
std      14.57738  14.57738    50.026628
min       0.00000   1.00000   233.600000
25%      12.25000  13.25000   253.900000
50%      24.50000  25.50000   286.500000
75%      36.75000  37.75000   317.775000
max      49.00000  50.00000   468.000000

На основі результатів методу `describe()`, можемо зробити наступні висновки:

1. **Rank**:
   - Маємо 50 значень (count), що підтверджує повноту даних.
   - Середній рейтинг (mean) — 24.5, а стандартне відхилення (std) — 14.58.
   - Мінімальне значення — 1, максимальне — 50, що підтверджує, що рейтинг є повним і послідовним.

2. **Cancer_Rate**:
   - Середній рівень захворюваності на рак — 294.08.
   - Мінімальний показник — 233.6, а максимальний — 468.0.
   - Значна варіативність рівня захворюваності, стандартне відхилення — 50.03, що вказує на великий розрив між країнами.

#### Unique

Метод `unique()` в Pandas використовується для виведення унікальних значень у стовпці даних. Він повертає масив усіх унікальних елементів, що містяться в обраному стовпці. Цей метод є корисним, коли потрібно дізнатися, які унікальні значення існують у категоріальному або числовому стовпці.

In [7]:
# Унікальні країни у стовпці 'Country'
unique_countries = cancer_data['Country'].unique()
print(unique_countries)

['Australia' 'New Zealand' 'Ireland' 'Hungary' 'United States' 'Belgium'
 'France' 'Denmark' 'Norway' 'Netherlands' 'Canada'
 'France (New Caledonia)' 'United Kingdom' 'South Korea' 'Germany'
 'Switzerland' 'Luxembourg' 'Serbia' 'Slovenia' 'Latvia' 'Slovakia'
 'Czech Republic' 'Sweden' 'Italy' 'Croatia' 'Lithuania' 'Estonia'
 'Greece' 'Spain' 'Finland' 'Uruguay' 'Belarus' 'Portugal' 'Iceland'
 'France (Guadeloupe)' 'United States (Puerto Rico)' 'Moldova' 'Poland'
 'Cyprus' 'France (Martinique)' 'Malta' 'Singapore' 'Japan' 'Austria'
 'Barbados' 'France (French Guiana)' 'Bulgaria' 'Lebanon'
 'France (French Polynesia)' 'Israel']


Аналізуючи унікальні значення в стовпці **Country**, можна побачити, що в наборі даних присутні 50 країн з унікальними назвами. Деякі країни мають кілька варіантів (наприклад, Франція з її заморськими територіями, такими як **France (Guadeloupe)**, **France (Martinique)**). Це свідчить про те, що різні регіони однієї країни можуть мати різні рівні захворюваності. Таким чином, унікальні значення допомагають ідентифікувати всі окремі регіони, що є важливим для подальшого аналізу.

#### Mean

Метод `mean()` в Pandas використовується для обчислення середнього значення числових стовпців. Він повертає середнє значення для всіх числових значень у вказаному стовпці. Це корисно для швидкого отримання середніх показників у наборі даних.


In [8]:
# Обчислюємо середнє значення рівня захворюваності на рак
average_cancer_rate = cancer_data['Cancer_Rate'].mean()
print(f"Середнє значення рівня захворюваності на рак: {average_cancer_rate}")

Середнє значення рівня захворюваності на рак: 294.08200000000005


Ми вже бачили середнє значення рівня захворюваності на рак, коли використовували метод `describe()`, і воно дорівнювало 294.08. Проте, ось ще раз це значення, обчислене за допомогою методу `mean()`, щоб показати середнє явно.

#### Виконаємо уявне завдання
Уявімо, що в нас не було стовпчику **Rank** і дані не були відсортовані. Метод `drop()` дозволяє видалити певний стовпчик або рядки з DataFrame. У нашому випадку ми видалили стовпчик **Rank**, щоб відновити його з нуля. Ми перемішали рядки за допомогою методу `sample()`, що випадково змінив порядок. Потім за допомогою методу `sort_values()` відсортували дані за стовпцем **Cancer_Rate** у порядку спадання. Після цього додали новий стовпчик **Rank**, який нумерує рядки.

In [10]:
# Видаляємо стовпчик Rank
data_without_rank = cancer_data.drop(columns=['Rank'])
shuffled_data = data_without_rank.sample(frac=1).reset_index(drop=True)

shuffled_data.head()

Unnamed: 0   Country  Cancer_Rate
0          18  Slovenia        304.9
1           6    France        344.1
2          40     Malta        249.4
3          26   Estonia        283.3
4          47   Lebanon        242.8

Як видно, в нас більше нема рейтингу, та дані ідуть не за порядком

In [12]:
sorted_data = shuffled_data.sort_values(by='Cancer_Rate', ascending=False).reset_index(drop=True)


sorted_data['Rank'] = range(1, len(sorted_data) + 1)

# Переглядаємо результат
sorted_data.head()

Unnamed: 0        Country  Cancer_Rate  Rank
0           0      Australia        468.0     1
1           1    New Zealand        438.1     2
2           2        Ireland        373.7     3
3           3        Hungary        368.1     4
4           4  United States        352.2     5

Колонка Rank тепер знову є, та ми змогли б зробити, навіть коли її не було спочатку.

#### Apply

Метод `apply()` використовується для застосування користувацької функції до кожного елемента або рядка DataFrame або Series. Він дозволяє виконувати перетворення над даними, які потребують логіки, яку не можна реалізувати за допомогою вбудованих функцій. 

Ви можете передати `apply()` власну функцію, яка виконує конкретне завдання — наприклад, категоризацію значень. Це дуже потужний інструмент для кастомної обробки даних.

Тепер застосуємо його для категоризації рівня захворюваності на рак.

In [16]:
# Функція для категоризації рівня захворюваності
def categorize_cancer_rate(rate):
    if rate > 350:
        return 'High'
    elif rate > 300:
        return 'Medium'
    else:
        return 'Low'

# Застосуємо функцію до стовпця 'Cancer_Rate'
cancer_data['Cancer_Rate_Category'] = cancer_data['Cancer_Rate'].apply(categorize_cancer_rate)

# Переглянемо результат
first_row = cancer_data.iloc[0]
middle_row = cancer_data.iloc[10]
last_row = cancer_data.iloc[-1]

# Виведемо результати
print("Перший рядок:\n", first_row)
print("\nСередній рядок:\n", middle_row)
print("\nОстанній рядок:\n", last_row)


Перший рядок:
 Unnamed: 0                      0
Rank                            1
Country                 Australia
Cancer_Rate                 468.0
Cancer_Rate_Category         High
Name: 0, dtype: object

Середній рядок:
 Unnamed: 0                  10
Rank                        11
Country                 Canada
Cancer_Rate              334.0
Cancer_Rate_Category    Medium
Name: 10, dtype: object

Останній рядок:
 Unnamed: 0                  49
Rank                        50
Country                 Israel
Cancer_Rate              233.6
Cancer_Rate_Category       Low
Name: 49, dtype: object


Як бачимо, в нас тепер є класи за рівнем захворюваності.

#### value_counts

Для підрахунку унікальних значень у новому стовпці **Cancer_Rate_Category**, можемо використовувати метод `value_counts()` для підрахунку кожної категорії.

Цей код виведе кількість унікальних значень (категорій) у стовпці, тобто скільки разів кожна категорія (High, Medium, Low) зустрічається в наборі даних.

In [17]:
# Підраховуємо унікальні значення в новому стовпчику 'Cancer_Rate_Category'
category_counts = cancer_data['Cancer_Rate_Category'].value_counts()

# Виводимо результат
print(category_counts)

Cancer_Rate_Category
Low       30
Medium    15
High       5
Name: count, dtype: int64


Як бачимо, за нашою класифікацією, в нас 5 країн з високим рівнем та 15 країн з середнім рівнем.

## Висновок

На цій лабораторній роботі я навчився вибирати відповідні дані з DataFrame, та вибирати правильні дані з набору даних. Також я вирішив кілька простих завдань за допомогою вивчених методів, та налаштував jupyter notebook для цього.